### DISCUSSION TITLE
Setting a global timeout on all steps

### DISCUSSION THREAD

**Discussion by ecaponi3** [cucumber-rs/cucumber#289] [Q&A] [answered]
We are encountering an internal error with our tests that cause a specific step to hang indefinitely. Im looking for methods to set a timeout value either on the step definition or on a global scale for all steps. Ive found mixed results online on whether or not there is a default timeout, but ideally if a test step is taking longer than lets say 30 seconds, we would like to just set it as a failed test and move on. 

Are there any cucumber RS methods for doing this?

Thank you

**Comment by tyranron** [comment, marked as answer]
@ecaponi3 hi! There is no such built-in functionality in `cucumber` crate. And I'm quite unsure, that `cucumber` crate is the correct place for it.

There are multiple ways of resolving this issue:

1. You can use [`tokio::time::timeout`](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html) (or similar), and wrap your step definitions into it:
    ```rust
    #[then("whatever")]
    async fn step(w: &mut World) {
        timeout(Duration::from_secs(30), async move {
            // your logic here...
        })
        .await
        .unwrap_or_else(|e| panic!("step timed out: {e}"))
    }
    ```
    With implementing a simple proc macro atop, it may be reduced to:
    ```rust
    #[then("whatever")]
    #[timeout(30)]
    async fn step(w: &mut World) {
        // your logic here...
    }
    ```

2. You can [implement your own `Runner`](https://cucumber-rs.github.io/cucumber/main/architecture/runner.html) which have this functionaly built-in, and use it in testing pipeline instead of the one provided by crate by default.

**Reply by ecaponi3** [reply]
@tyranron, awesome thank you for the suggestion. The step that we are seeing get stuck (indefinitely) is a setup step (when), that is used in nearly all of our feature files, so it makes sense to just implement the Tokio suggestion directly on that step and see if we can solve our issues that way. 

I will try this and report back if it solves the issues we are having. 

Thanks again
